datumaro/components/media.py

```diff
@@ -178,3 +179,255 @@ def save(self, path):
+        return self._video.frame_size
+
+    @property
+    def index(self) -> int:
+        return self._index
+
+    @property
+    def video(self) -> 'Video':
+        return self._video
+
+class _VideoFrameIterator(Iterator[VideoFrame]):
+    """
+    Provides sequential access to the video frames.
+    """
+
+    def __init__(self, video: 'Video'):
+        self._video: 'Video' = video
+        self._iterator: Iterator[VideoFrame] = None
+        self._pos: int = -1
+        self._current_frame_data: Optional[np.ndarray] = None
+
+        self._reset()
+
+    def _reset(self):
+        self._video._reset()
+        self._iterator = self._decode(self._video._get_container())
+        self._pos = -1
+        self._current_frame_data = None
+
+    def _decode(self, cap) -> Iterator[VideoFrame]:
+        """
+        Decodes video frames using opencv
+        """
+
+        self._pos = -1
+
+        success, frame = cap.read()
+        while success:
+            self._pos += 1
+            if self._video._includes_frame(self._pos):
+                self._current_frame_data = frame.astype(float)
+                yield VideoFrame(self, index=self._pos)
+
+            success, frame = cap.read()
+
+        if self._video._frame_count is None:
+            self._video._frame_count = self._pos + 1
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return next(self._iterator)
+
+    def __getitem__(self, idx: int) -> VideoFrame:
+        if not self._video._includes_frame(idx):
+            raise IndexError(f"Video doesn't contain frame #{idx}.")
+
+        return self._navigate_to(idx)
+
+    def get_frame_data(self, idx: int) -> np.ndarray:
+        self._navigate_to(idx)
+        return self._current_frame_data
+
+    def _navigate_to(self, idx: int) -> VideoFrame:
+        """
+        Iterates over frames to the required position.
+        """
+
+        if idx < 0:
+            raise IndexError()
+
+        if idx < self._pos:
+            self._reset()
+
+        if self._pos < idx:
+            try:
+                while self._pos < idx:
+                    v = self.__next__()
+            except StopIteration as e:
+                raise IndexError() from e
+        else:
+            v = VideoFrame(video=self, index=self._pos)
+
+        return v
+
+class Video(MediaElement, Iterable[VideoFrame]):
+    """
+    Provides random access to the video frames.
+    """
+
+    def __init__(self, path: str, *,
+            step: int = 1, start_frame: int = 0,
+            end_frame: Optional[int] = None) -> None:
+        super().__init__(path)
+
+        self._iterator = None
+        self._frame_size: Optional[Tuple[int, int]] = None
+
+        self._container = None
+        self._reset()
+
+        self._step = int(step)
+        self._start_frame = int(start_frame)
+        self._end_frame = int(end_frame) if end_frame else None
+
+        self._frame_count = self._get_frame_count(self._container)
```

Comment:
Do both `self._end_frame` and `self._frame_count` need to exist? It seems like you don't really need both of them after the initialization is complete. I think the following initialization algorithm would be sufficient:

1. Try to fetch the total frame count for the video (I'll call this `frame_count`).
1. If `end_frame` is `None`, set `end_frame = frame_count`. Otherwise, verify that `frame_count is None or end_frame <= frame_count`.
1. If `end_frame` is not `None`, verify that `end_frame > start_frame`.
1. Save `self._start_frame = start_frame; self._end_frame = end_frame`. Now you can always use `self._end_frame` as the end of the segment.

Reply:
Unfortunately, we need to initialize `frame_count` lazily, because not all videos can provide frame count info. Honestly, I'm not even sure we should use it all, because the reported value can be different from the real frame count (because FPS can be variable and frame count is an estimation obtained as `fps * duration`), however it can be a significant performance issue if we try to decode all the frames just to count them. So the idea is that we just save the input values and defer the action you suggested to the `get_end_frame` method, which always checks if we already know the real frame count. The bad thing about this is that `Video` can report different frame count after we get the real value. A dedicated metafile describing frames could be useful here. I'll think what we can do here.